segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/SegmentAggregator.java

```diff
@@ -1516,6 +1523,11 @@ private void checkValidOperation(StorageOperation operation) throws DataCorrupti
         if (this.hasSealPending.get() && !isTruncateOperation(operation) && !isDeleteOperation(operation)) {
             throw new DataCorruptionException(String.format("Illegal operation for a sealed Segment; received '%s'.", operation));
         }
+
+        // Check that operations being processed follow a monotonically increasing order.
+        Preconditions.checkArgument(operation.getSequenceNumber() > this.lastAddedSequenceNumber.get(),
+                "Operation %s for Segment %s does not follow monotonically increasing order.", operation, operation.getStreamSegmentId());
+        this.lastAddedSequenceNumber.set(operation.getSequenceNumber());
```

Comment:
This can be setAndGet and then assert on the result so that it is atomic.

Reply:
The reason why I did not do that is that, if the precondition fails, the `this.lastAddedSequenceNumber` may get incremented anyway, right? I we don't want this to happen, only incrementing that value once the precondition passes.